File: Algo JS/binarytreeprinter.js
/***********************************************************************
 * Binary Tree Printer class - prints binary tree 
 ***********************************************************************/

// import modules
const Color = require('./color');
const colors = require('colors');
const ConsoleProgressBar = require('./consoleprogressbar');

module.exports = class BinaryTreePrinter
{
    static get Legend() {
        return Legend;
    }
    
    static printTopToBottom(tree) {
        const LEFT_BRANCH = "┌";
        const RIGHT_BRANCH = "┐";
        const BRANCH = "|";
        const DEFAULT_SPACING = 4;
        
        throw Error("unsupported function!");
        
    }
    
    static printLeftToRight(tree) {
        const RIGHT_BRANCH = "┌── ";
        const LEFT_BRANCH = "└── ";
        const DEFAULT_SPACING = 4;
        
        let _printBar = [];
        const _tree = tree;
        let numNodesProcessed = 0;
        let progressBar = null;
        
        for (let index = 0; index < tree.height; index++) {
            _printBar[index] = false;
        }
        
        _print(tree);
        
        function _print(tree) {
            if (tree.isEmpty()) {
                console.log("[ EMPTY TREE ]");  
                return;
            }
            
            let source = tree.root;
            let stringBuilder = "\n\n";
            
            //progressBar = new ConsoleProgressBar();
            progressBar = new ConsoleProgressBar(tree.size);
            progressBar.start();

            stringBuilder = _printNode(source, stringBuilder);
            
            progressBar.stop();
            
            console.log(stringBuilder);
            console.log("\n\n");
        }
        
        
        // inorder tree walk -- right child first
        function _printNode(node, stringBuilder) {
            if (node !== null) 
            {
                numNodesProcessed += 1;
                progressBar.update(numNodesProcessed);
                
                // print right child first
                stringBuilder = _printNode(node.right, stringBuilder);

                // then print node
                if (node.isRoot()) {
                    stringBuilder += ("" + node.key).black.bold;
                    stringBuilder += "──|\n";
                }
                
                else {
                    stringBuilder = _pad(node, stringBuilder);
                    
                    if (node.color === Color.DOUBLEBLACK) {
                        stringBuilder += ("" + node.key).inverse;
                    }
                    else if (node.color === Color.RED) {
                        stringBuilder += ("" + node.key).red.bold;
                    }
                    else {
                        stringBuilder += ("" + node.key).black.bold;
                    }
                    
                    stringBuilder += "\n";
                }

                // keep track where to print bar | while drawing tree
                if (node.parent !== null && node.isRightChild()) {
                    let parentLevel = _tree.getLevel(node.parent);
                    _printBar[parentLevel] = true;
                }
                
                else if (node.parent !== null && node.isLeftChild()) {
                    let parentLevel = _tree.getLevel(node.parent);
                    _printBar[parentLevel] = false;
                }
                
                if (!node.hasLeft()) {
                    _printBar[_tree.getLevel(node)] = false;
                }
                
                // print left child last
                stringBuilder = _printNode(node.left, stringBuilder);
                
            }
            
            return stringBuilder;
        }
        
        function _pad(node, stringBuilder) {
            let level = _tree.getLevel(node);
            
            for (let i = 0; i < level; i++) {
                stringBuilder += "".padStart(DEFAULT_SPACING); 
                
                if (_printBar[i] === true) 
                {
                    // don't need a bar before └── 
                    if ((node.isLeftChild()) && (i === (level-1))) {
                        stringBuilder += "";
                    }
                    else {
                        // replace last space with |
                        stringBuilder = stringBuilder.slice(0, stringBuilder.length - 1);
                        stringBuilder += "|";
                    }
                }
            }
            
            if (node.hasParent()) 
            {
                // remove last space to accomodate └── or ┌── 
                //if (stringBuilder[stringBuilder.length - 1] !== "\n") 
                {
                    stringBuilder = stringBuilder.slice(0, stringBuilder.length - 1);
                }
                        
                if (node.isLeftChild()) {
                    stringBuilder += LEFT_BRANCH;
                }
                else if (node.isRightChild()) {
                    stringBuilder += RIGHT_BRANCH;
                }
            }  
                
            return stringBuilder;
        }
    }
    
    static printLinuxStyle(tree, legend = Legend.NO) {
        
        const HALF_BRANCH = "└── ";
        const FULL_BRANCH = "├── ";
        const DEFAULT_SPACING = 4;
        let _printBar = [];
        const _tree = tree;
        
        for (let index = 0; index < tree.height; index++) {
            _printBar[index] = true;
        }
        
        _print(tree);
        
        // prints binary tree from root to leaf, left to right
        function _print(tree) {
            if (tree.root === null) {
                console.log("[ EMPTY TREE ]");
                return;
            }

            let stringBuilder = "\n";
            let source = tree.root;
            
            if (legend === Legend.YES) {
                stringBuilder = _printLegend(stringBuilder);
            }
            
            stringBuilder += HALF_BRANCH;
            stringBuilder = _printNode(source, stringBuilder);

            console.log(stringBuilder.toString());
        }
        
        // recursively print node -- preorder tree walk
        function _printNode(node, stringBuilder) {

            // print the node
            if (node.color === Color.DOUBLEBLACK) {
                stringBuilder += ("" + node.key).black.bold;
            }
            else if (node.color === Color.RED) {
                stringBuilder += ("" + node.key).red.bold;
            }
            else  {
                stringBuilder += ("" + node.key).black.bold;
            }
            
            stringBuilder += "\n";

            
            if (node.parent !== null && node.isRightChild()) {
                let parentLevel = _tree.getLevel(node.parent);
                
                _printBar[parentLevel] = false;
            }
            else if (node.parent !== null && node.isLeftChild()) {
                let parentLevel = _tree.getLevel(node.parent);
                
                _printBar[parentLevel] = true;
            }

            if (node.hasLeft() || node.hasRight()) {
                stringBuilder = _pad(node, stringBuilder);
            }

            if (node.hasLeft() && node.hasRight()) {
                stringBuilder += FULL_BRANCH;
                stringBuilder = _printNode(node.left, stringBuilder);


                stringBuilder = _pad(node, stringBuilder);
                stringBuilder += "|\n";
                stringBuilder = _pad(node, stringBuilder);
                
                stringBuilder += HALF_BRANCH;
                stringBuilder = _printNode(node.right, stringBuilder);
            }
            else {
                if (node.hasLeft()) {
                    stringBuilder += HALF_BRANCH;
                    stringBuilder = _printNode(node.left, stringBuilder);
                }
                else if (node.hasRight()) {
                    stringBuilder += "|";
                    stringBuilder += "\n";
                    stringBuilder = _pad(node, stringBuilder);
                    stringBuilder += "|";
                    stringBuilder += "\n";
                    stringBuilder = _pad(node, stringBuilder);
                    stringBuilder += HALF_BRANCH;
                    stringBuilder = _printNode(node.right, stringBuilder);
                }
            }

            return stringBuilder;
        }

        function _pad(node, stringBuilder) {
            let level = _tree.getLevel(node);
            
            stringBuilder += "".padStart(DEFAULT_SPACING);
            
            for (let i = 0; i < level; i++) {
                if (_printBar[i] === true) 
                {
                    stringBuilder += "|";
                    stringBuilder += "".padStart(DEFAULT_SPACING - 1);
                }
                else {
                    stringBuilder += "".padStart(DEFAULT_SPACING);
                }
            }

            return stringBuilder;
        }
        
        function _printLegend(stringBuilder) {
            stringBuilder += "LEGENDS: \n";
            stringBuilder += ("        " + HALF_BRANCH + "node\n");
            stringBuilder += ("                " + FULL_BRANCH + "left.child\n");
            stringBuilder += ("                |             |\n");
            stringBuilder += ("                |             |\n");
            stringBuilder += ("                |             " + HALF_BRANCH + "right.child\n");
            stringBuilder += "                |\n";
            stringBuilder += "                |\n";
            stringBuilder += ("                " + HALF_BRANCH + "right.child\n");
            stringBuilder += ("                               " + HALF_BRANCH + "left.child\n");
            stringBuilder += "\n\n";
            
            return stringBuilder;
        }
    }
    
};

const Legend = Object.freeze({
    YES : "YES",
    NO : "NO"
});

File: Algo Java/binarysearchtrees/BinaryTreePrinter.java
/**
 *
 * @author Gaurab R. Gautam
 */
package binarysearchtrees;

import utility.ConsoleProgressBar;
import utility.OSDetector;



public class BinaryTreePrinter {
    static final String RIGHT_BRANCH = "┌── ";
    static final String LEFT_BRANCH = "└── ";
    static final String VERTICAL_LINE = "\u2502";
    static final int DEFAULT_SPACING = 6; 
    static int numNodesProcessed = 1;
    static ConsoleProgressBar progressBar;
        
    private static String red(String text)
    {
        final String ANSI_RESET = "\u001B[0m";
        final String ANSI_RED = "\u001B[31m";
        
//        if (OSDetector.getOS() == OSDetector.OS.WINDOWS)
//        {
//            return text;
//        }
//        else
//        {
//            return ANSI_RED + text + ANSI_RESET;
//        }
        
        return ANSI_RED + text + ANSI_RESET;
    }
    
    private static String black(String text)
    {
        final String ANSI_RESET = "\u001B[0m";
        final String ANSI_BLACK = "\u001B[30m";
        
//        if (OSDetector.getOS() == OSDetector.OS.WINDOWS)
//        {
//            return text;
//        }
//        else
//        {
//            return ANSI_BLACK + text + ANSI_RESET;
//        }
        
        return ANSI_BLACK + text + ANSI_RESET;
    }
    
    private static String doubleBlack(String text)
    {
        final String ANSI_RESET = "\u001B[0m";
        final String ANSI_WHITE_BACKGROUND = "\u001B[47m";
        final String ANSI_BOLD = "\u001B[1m";
        final String ANSI_BLACK = "\u001B[30m";
        
//        if (OSDetector.getOS() == OSDetector.OS.WINDOWS)
//        {
//            return text;
//        }
//        else
//        {
//            return ANSI_BOLD + ANSI_WHITE_BACKGROUND + ANSI_BLACK  + text + ANSI_RESET;  
//        }

        return ANSI_BOLD + ANSI_WHITE_BACKGROUND + ANSI_BLACK  + text + ANSI_RESET;  
    }   
    
    public static void printLeftToRight(BinarySearchTree tree) {
        
        boolean[] _printBar = new boolean [tree.getHeight()];
        
        for (int index = 0; index < tree.getHeight(); index++) {
            _printBar[index] = false;
            
        }
        
        print(tree, _printBar);
        
    }
        
    private static void print(BinarySearchTree tree, boolean[] _printBar) 
    {
        if (tree.isEmpty()) {
            System.out.println("[ EMPTY TREE ]");  
            return;
        }

        Node source = tree.root;
        String stringBuilder = "\n\n";
        
        progressBar = new ConsoleProgressBar(tree.getSize());
        progressBar.setHeader("Printing");
        progressBar.setProgressBarIndicator(ConsoleProgressBar.ProgressBarIndicator.HALF_BLOCK);
        progressBar.start();
        
        
        stringBuilder = printNode(tree, source, stringBuilder, _printBar);
        
        progressBar.stop();
        System.out.println(stringBuilder);
        System.out.println("\n\n");
    }
        
    // inorder tree walk -- right child first
    private static String printNode(BinarySearchTree tree, Node node, 
                                        String stringBuilder, boolean[] _printBar) 
    {
        if (node != null) 
        {
            // print right child first
            stringBuilder = printNode(tree, node.right, stringBuilder, _printBar);

            // then print node
            if (node.isRoot()) {
                stringBuilder += ("" + node.key);
                stringBuilder += "──|\n";
            }

            else {
                stringBuilder = _pad(tree, node, stringBuilder, _printBar);
                
                if (node.color == null)
                {
                    stringBuilder += black(("" + node.key));
                }
                else 
                {
                    switch (node.color) 
                    {
                        case DOUBLEBLACK:
                            stringBuilder += doubleBlack(("" + node.key));
                            break;
                        case RED:
                            stringBuilder += red(("" + node.key));
                            break;
                        case BLACK:
                            stringBuilder += black(("" + node.key));
                            break;
                    }
                }

                stringBuilder += "\n";
            }
            
            // update progress of node processed
            numNodesProcessed += 1;
            progressBar.update(numNodesProcessed);
            

            // keep track where to print bar | while drawing tree
            if (node.parent != null && node.isRightChild()) {
                int parentLevel = tree.getLevel(node.parent);
                _printBar[parentLevel] = true;
            }

            else if (node.parent != null && node.isLeftChild()) {
                int parentLevel = tree.getLevel(node.parent);
                _printBar[parentLevel] = false;
            }

            if (!node.hasLeft()) {
                _printBar[tree.getLevel(node)] = false;
            }

            // print left child last
            stringBuilder = printNode(tree, node.left, stringBuilder, _printBar);

        }

        return stringBuilder;
    }
        
    private static String _pad(BinarySearchTree tree, Node node, String stringBuilder, boolean[] _printBar) {
        int level = tree.getLevel(node);
        
        for (int i = 0; i < level; i++) {
            
            stringBuilder += String.format("%" + DEFAULT_SPACING + "s", ""); 

            if (_printBar[i] == true) 
            {
                // don't need a bar before └── 
                if ((node.isLeftChild()) && (i == (level-1))) {
                    stringBuilder += "";
                }
                else {
                    // replace last space with |
                    stringBuilder = stringBuilder.substring(0, stringBuilder.length() - 1);
                    stringBuilder += VERTICAL_LINE;
                }
            }
        }

        if (node.hasParent()) 
        {
            // remove last space to accomodate └── or ┌── 
            //if (stringBuilder[stringBuilder.length - 1] !== "\n") 
            {
                stringBuilder = stringBuilder.substring(0, stringBuilder.length() - 1);
            }

            if (node.isLeftChild()) {
                stringBuilder += LEFT_BRANCH;
            }
            else if (node.isRightChild()) {
                stringBuilder += RIGHT_BRANCH;
            }
        }  

        return stringBuilder;
    }
}
